components/Msig/ApproveCancel.tsx

```diff
@@ -0,0 +1,156 @@
+  const [txFee, setTxFee] = useState<FilecoinNumber | null>(null)
+
+  // Get transaction info from url
+  const { proposal } = router.query
+  const transaction = useMemo<MsigTxWithApprovals | null>(() => {
+    try {
+      return JSON.parse(decodeURI(proposal as string))
+    } catch (e) {
+      setTxState(TxState.LoadingFailed)
+      return null
+    }
+  }, [proposal])
+
+  // Create message from input
+  const message = useMemo<Message | null>(
+    () =>
+      transaction
+        ? new Message({
+            to: Address,
+            from: wallet.address,
+            nonce: 0,
+            value: 0,
+            method,
+            params: Buffer.from(
+              serializeParams({
+                TxnID: transaction.id,
+                ProposalHashData: transaction.proposalHash
+              }),
+              'hex'
+            ).toString('base64'),
+            gasPremium: 0,
+            gasFeeCap: 0,
+            gasLimit: 0
+          })
+        : null,
+    [Address, wallet.address, method, transaction, serializeParams]
+  )
+
+  // Get actor data from transaction
+  const { data: actorData, error: actorError } = useActorQuery({
+    variables: { address: convertAddrToPrefix(transaction?.to.robust) },
+    skip: !transaction?.to.robust
+  })
+
+  // Get actor name from actor data
+  const actorName = useMemo<string>(
+    () => (actorData ? decodeActorCID(actorData.actor.Code) : ''),
+    [actorData]
+  )
+
+  // Set TxState.FillingForm when actor name has loaded
+  useEffect(() => actorName && setTxState(TxState.FillingForm), [actorName])
+
+  // Set TxState.LoadingFailed when actor name failed to load
+  useEffect(() => actorError && setTxState(TxState.LoadingFailed), [actorError])
+
+  return (
+    <Transaction.Form
+      title={title}
+      description={description}
+      msig
+      method={method}
+      message={message}
+      total={txFee}
+      txState={txState}
+      setTxState={setTxState}
+      maxFee={wallet.balance}
+      txFee={txFee}
+      setTxFee={setTxFee}
+      onComplete={() => navigate(router, { pageUrl: PAGE.MSIG_HISTORY })}
+      walletProviderOpts={walletProviderOpts}
+      pendingMsgContext={pendingMsgContext}
+    >
+      <Transaction.Balance
+        address={wallet.address}
+        balance={wallet.balance}
+        msigBalance={AvailableBalance}
+      />
+      {transaction && actorName && (
+        <Parameters params={transaction} depth={0} actorName={actorName} />
+      )}
+    </Transaction.Form>
+  )
+}
+
+export enum ApproveCancelMode {
+  APPROVE = 'APPROVE',
+  CANCEL = 'CANCEL'
+}
+
+const APPROVE_CANCEL_MODE_PROPTYPE = PropTypes.oneOf(
+  Object.values(ApproveCancelMode) as Array<ApproveCancelMode>
+)
+
+interface ApproveCancelProps {
+  mode: ApproveCancelMode
+  walletProviderOpts?: WalletProviderOpts
+  pendingMsgContext?: Context<PendingMsgContextType>
+}
+
+ApproveCancel.propTypes = {
+  mode: APPROVE_CANCEL_MODE_PROPTYPE.isRequired,
```

Comment:
I think this should use the MsigMethod type (see comments below for link to it). Just less variables